discovery/cvm/cvm.go

```diff
@@ -0,0 +1,292 @@
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	d := &Discovery{
+		conf:   conf,
+		logger: logger,
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"cvm",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	_ = level.Debug(d.logger).Log("refresh with sd conf:", d.conf.Region, d.conf.Endpoint, d.conf.Filters)
+
+	client, err := getCvmClient(d.conf)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create cvm client")
+	}
+	tg := &targetgroup.Group{
+		Source: d.conf.Region,
+	}
+
+	var filters []*cvm.Filter
+	for _, f := range d.conf.Filters {
+		filters = append(filters, &cvm.Filter{
+			Name:   common.StringPtr(f.Name),
+			Values: common.StringPtrs(f.Values),
+		})
+	}
+	request := cvm.NewDescribeInstancesRequest()
+	request.Filters = filters
+	request.Limit = common.Int64Ptr(DefaultPageLimit)
+
+	var offset int64 = 0
+
+	for {
+		request.Offset = common.Int64Ptr(offset)
+		_ = level.Debug(d.logger).Log("retrieving instances, offset:", *request.Offset)
+
+		resp, err := client.DescribeInstances(request)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not describe cvm instances")
+		}
+
+		for _, inst := range resp.Response.InstanceSet {
+			if len(inst.PrivateIpAddresses) == 0 {
+				continue
+			}
+			tg.Targets = append(tg.Targets, d.extractLabels(inst))
+		}
+
+		if len(resp.Response.InstanceSet) < DefaultPageLimit {
+			break
+		}
+		offset += DefaultPageLimit
+	}
+	return []*targetgroup.Group{tg}, nil
+}
+
+func (d *Discovery) extractLabels(inst *cvm.Instance) model.LabelSet {
+	labels := model.LabelSet{
+		cvmLabelInstanceID: strPtrToLabelValue(inst.InstanceId),
+		cvmLabelRegion:     model.LabelValue(d.conf.Region),
+	}
+
+	addr := net.JoinHostPort(*inst.PrivateIpAddresses[0], fmt.Sprintf("%d", d.conf.Port))
+	labels[model.AddressLabel] = model.LabelValue(addr)
+
+	if inst.Placement != nil {
+		labels[cvmLabelZone] = strPtrToLabelValue(inst.Placement.Zone)
+		labels[cvmLabelProjectId] = intPrtToLabelValue(inst.Placement.ProjectId)
+		if inst.Placement.HostIds != nil {
+			labels[cvmLabelHostId] = strPtrSliceToLabelValue(inst.Placement.HostIds, false)
+		}
+	}
+
+	labels[cvmLabelInstanceType] = strPtrToLabelValue(inst.InstanceType)
+	labels[cvmLabelPrivateIP] = strPtrSliceToLabelValue(inst.PrivateIpAddresses, false)
+	labels[cvmLabelPublicIP] = strPtrSliceToLabelValue(inst.PublicIpAddresses, false)
+	labels[cvmLabelInstanceState] = strPtrToLabelValue(inst.InstanceState)
+
+	if inst.VirtualPrivateCloud != nil {
+		labels[cvmLabelVPCID] = strPtrToLabelValue(inst.VirtualPrivateCloud.VpcId)
+		labels[cvmLabelSubnetID] = strPtrToLabelValue(inst.VirtualPrivateCloud.SubnetId)
+	}
+
+	for _, tag := range inst.Tags {
+		labelName := model.LabelName(*tag.Key)
+		labels[cvmLabelTag+labelName] = strPtrToLabelValue(tag.Value)
+		// if key is not valid as label name, encode it to keep all information
```

Comment:
here we did not use strutil.SanitizeLabelName on purpose, because in our production environment, we sometimes have tag.key which contains chinese or japanese, use strutil.SanitizeLabelName will lose information, so we just check labelName is valid or not, if not valid, we encode tag.key as a valid labelName